### ISSUE TITLE
Android2D's performance is low

### ISSUE THREAD

**Issue by codeanticode** [processing/processing-android#265]
A full-screen app on a N5X with fairly simple graphics runs at ~3fps, while P2D gives 60fps

**Comment by benfry** [comment]
What happened? Even the original G1 was never that bad.

**Comment by codeanticode** [comment]
Apart from the modifications in the core library's architecture (PSurface, etc) nothing changed in PGraphicsAndroid2D, and the new architecture should not affect how the renderer handles the canvas. But I may be missing something, so will be looking into it.

**Comment by victordiaz** [comment]
I experienced the same.
The screen the resolution in newer phones are way higher than old ones, and people tend to play with full resolution so each redraw is just too much for hires phones. Just compare the newer Pixel phone 1,440x2,560 vs the first HTC Magic / G1 480x320

Would it make sense to add a scaling factor in the Android version? using a downsized canvas and upscale it when drawing.

**Comment by codeanticode** [comment]
@victordiaz Looks like the problem is that the offscreen canvas that PGraphicsAndroid2D uses for drawing is not hardware accelerated, see this [stackoverflow thread](http://stackoverflow.com/questions/9588776/canvas-in-surfaceview-hardware-acceleration) where one of the Android software engineers confirms that is was still the case for 4.2.1. I tested on devices with Android 5 and 7, and canvas is not hardware accelerated yet.

According to [Google's documentation](https://developer.android.com/guide/topics/graphics/hardware-accel.html), "Android 2D rendering pipeline supports hardware acceleration, meaning that all drawing operations that are performed on a View's canvas use the GPU" and that "Hardware acceleration is enabled by default if Target API level is >=14", but this seems to apply only if one uses the Canvas object provided in the View.onDraw(Canvas) callback (see [this](http://stackoverflow.com/questions/14785060/android-canvas-and-hardware-acceleration) and [this](http://stackoverflow.com/questions/19852820/can-we-hardware-accelerate-an-off-screen-canvas)). Even in that case, using a hardware-accelerated canvas could be problematic, as not all drawing functions are [supported](https://developer.android.com/guide/topics/graphics/hardware-accel.html#unsupported) in older API levels.

Of course, you can always use the OpenGL renderers (P2D/P3D), which are fully hardware-accelerated and should be reasonable close to the default renderer in quality (specially if using smooth()).

In principle, it would be possible to implement your idea of drawing to a downsized canvas and upscaling when rendering it to the screen... in fact, part of the scaling logic is [already there](https://github.com/processing/processing-android/blob/master/core/src/processing/core/PImage.java#L58) (at least to keep compatibility with the Java mode), but we will need to add some new API to run the sketch in fullscreen mode with a downscaled canvas resolution, which I'm not sure it is worth the effort.

**Comment by victordiaz** [comment]
@codeanticode Yes, you are right about the lack of hardware acceleration in many methods of the Android Canvas.
On the other hand I would consider having a scaling mode, not only for the canvas mode, but for the opengl mode. I've tried many lowcost / old devices where Processing doesnt run smoothly, specially tablets. (higres screen and not performant gpu).

One little trick I've been using was scaling the drawing View using android methods (view.setScaleX/Y). Easy to implement although it might have some drawbacks depending on how Processing for Android is implemented.

**Comment by loooog** [comment]
Extremely low ,too slow .any way to compile to. Ndk or CPP stuff

**Comment by codeanticode** [comment]
We are looking into this, considering [this contribution](https://github.com/processing/processing/issues/5256) for faster P2D on desktop as a possibility.

**Comment by codeanticode** [comment]
Testing the new P2X renderer [in this branch](https://github.com/processing/processing-android/tree/super-fast-P2D). Need to integrate it better with the existing code, but is looking good. [Version 4.1](https://github.com/processing/processing-android/milestone/7) will be an exciting release of the mode!

**Comment by codeanticode** [comment]
Fixed a number of issues in the experimental P2DX renderer included in 4.1.0-beta3, depth algorithm works with a 16-bit buffer, and custom shaders are supported although they require different uniforms than regular Processing shaders.

**Comment by hazmatsuitor** [comment]
Sorry to have abandoned this code. I was changing a lot of what I was doing with programming around the time I finished writing it, and haven't really done Java programming since. It's good to see it finding some use, and thanks for fixing up some of the rough edges.

Reading back through the code, I'm not really happy with some things. The line rendering code in particular could be made faster by using linear algebra alternatives to the trigonometry it currently uses. I'll see if I can improve that.

**Comment by hazmatsuitor** [comment]
@codeanticode https://gist.github.com/notnullnotvoid/667bd229cc4ab5c93cbc7b1ae21e23c9 here I've written faster versions of several things. The most significant change is that all the code related to rendering circles and arcs now uses a rotation matrix so that `sin()` and `cos()` only have to be called one time at the beginning rather than for every point. I've also rewritten the line rendering code so it no longer uses any trig functions at all for miter and bevel joints, which actually simplified the code significantly. The shape and line rendering benchmarks are in the ballpark of ~1.5x to 2.0x faster (as measured on my machine) depending one which shapes are drawn and which line cap/joint settings are used. I also made a minor tweak to `postMatrixChanged()` to call `sqrt()` only once instead of twice. If the contents of `postMatrixChanged()` get folded into `circleDetail()` like I suggested in the original issue thread, then that should help quite a bit.

**Comment by codeanticode** [comment]
@hazmatsuitor great, thanks so much for picking this thread up again, and making this improvements! I will incorporate them very soon. Hopefully I can backport all of this to the Java mode in a not-so-distant future :-)

**Comment by codeanticode** [comment]
@hazmatsuitor I incorporated your latest code with https://github.com/processing/processing-android/commit/757e03bf74ff4c1a174702fb2695835f17a6ff18

As far as the original issue, I think that the goal of improving 2D rendering performance has been reasonably achieved. The remaining work consists in integrating P2DX with the existing shaders, as I pointed out earlier. So I'm going to close this issue, and open a new one specifically about that: #565

Thanks again for this amazing contribution!
